reconcile/utils/oc.py

```diff
@@ -63,6 +64,55 @@ class JobNotRunningError(Exception):
 
 
 class OC:
+    class _Decorators:
+        @classmethod
+        def process_reconcile_time(cls, function):
+
+            @wraps(function)
+            def wrapper(*args, **kwargs):
+                result = function(*args, **kwargs)
+
+                running_state = RunningState()
+                commit_time = float(running_state.timestamp)
+                time_spent = time.time() - commit_time
+
+                try:
+                    kind = result['resource']['kind']
+                    name = result['resource']['metadata']['name']
+                    annotations = \
+                        result['resource']['metadata'].get('annotations')
+                except KeyError as e:
+                    logging.warning(f"Error processing metric: {e}")
+                    return
+
+                ignore_reconcile_time = \
+                    annotations.get('qontract.ignore_reconcile_time') == 'true'
+                if not ignore_reconcile_time:
+                    name = f'{function.__module__}.{function.__qualname__}'
+                    reconcile_time.labels(
+                        name=name,
+                        integration=running_state.integration
+                    ).observe(amount=time_spent)
+
+                if not result['is_log_slow_oc_reconcile']:
+                    return
+
+                if time_spent > result['slow_oc_reconcile_threshold']:
+                    msg = f"{kind} {name} in namespace " \
+                          f"{result['namespace']} from " \
+                          f"{result['server']} took {time_spent} to " \
+                          f"reconcile. Commit sha {running_state.commit} " \
+                          f"and commit ts {running_state.timestamp}."
+
+                    if ignore_reconcile_time:
+                        msg += " Ignored in the metric published."
+
+                    logging.info(msg)
+
+                return
+
```

Comment:
so, my comments here are around making a safe decorator. there are 2 ways to approach that: testing each argument (with try/except or with get()/if). Or, the way I preffer in this case, by having the `function(*args, **kwargs)` returning a custom class instance trusted to have all you need, then you don't have to check any arguments in advance, but you do have to check if the function returned the object you need.

Reply:
I ok, with both, just wanted to raise the topic.

Reply:
So since everything comes from `_msg_to_process_reconcile_time` I was trusting I had everything I can. But I understand where you come from and it makes sense. If we want to make this safe I like the custom class approach better.

Reply:
Fixed in 69576f7